smt.go

```diff
@@ -1,210 +1,392 @@
 // Package smt implements a Sparse Merkle tree.
 package smt
 
-import(
-    "hash"
+import (
+	"bytes"
+	"hash"
 )
 
 const left = 0
 const right = 1
-var defaultValue = []byte{0}
+
+var defaultValue = []byte{}
+
+type keyAlreadyEmptyError struct{}
+
+func (e *keyAlreadyEmptyError) Error() string {
+	return "key already empty"
+}
```

Comment:
How is this error useful if not public? If it is public users can switch over it and handle it accordingly.  
Ah OK, I see: this is only used internally (and switched over). 

This can be simplified to `var keyAlreadyEmptyError = errors.New("key already empty")` , then you can do 
`if err == keyAlreadyEmptyError ` instead of `if _, ok := err.(*keyAlreadyEmptyError); ok {`.